parser.cc

```diff
@@ -1453,6 +1454,52 @@ Input::Index AsciiFileInput::size() const
 	return file_size;
 }
 
+StreamInput StreamInput::Create(std::istream& s)
+{
+	const size_t start = s.tellg();
+	s.seekg(0, std::ios::end);
+
+	const size_t len = static_cast<size_t>(s.tellg()) - start;
+	s.seekg(start, std::ios::beg);
+
+	return StreamInput(s, len);
+}
+
+StreamInput::StreamInput(std::istream& s, size_t len)
+	: length(len), stream(s)
+{
+}
+
+bool StreamInput::fillBuffer(Index start, Index &len, char32_t *&b)
+{
+	if (start > length)
+	{
+		return false;
+	}
+
+	char buffer[static_buffer_size];
+	len = std::min(this->length, static_buffer_size);
+	len = std::min(len, this->length - start);
+
+	stream.read(buffer, len);
```

Comment:
Ah, good point.  We could make this a template, but that exposes a bit more than I'm comfortable with in the header, so this os probably fine as long as there's a note in the constructor that this will move the stream position around and that it's undefined to try to use the stream elsewhere.

I don't really like it, because that then makes working out when to close the stream a bit difficult.